File: Go/030. Substring with Concatenation of All Words.go
package main

func findSubstring(s string, words []string) []int {
	res, str2num := make([]int, 0), make(map[string]int)
	if len(s) == 0 || len(words) == 0 || len(words[0]) == 0 {
		return res
	}
	for _, str := range words {
		if _, ok := str2num[str]; !ok {
			str2num[str] = 1
		} else {
			str2num[str] += 1
		}
	}
	for i := 0; i <= len(s) - len(words) * len(words[0]); i++ {
		seen, tag := make(map[string]int), true
		for k, v := range str2num {
			seen[k] = v
		}
		for j, n := 0, len(words[0]); tag && j < len(words); j++ {
			str := s[i + j * n : i + (j + 1) * n]
			if val, ok := seen[str]; ok && val > 0 {
				seen[str]--
			} else {
				tag = false
			}
		}
		if tag {
			res = append(res, i)
		}
	}
	return res
}

File: C++/030. Substring with Concatenation of All Words.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> findSubstring(std::string s, std::vector<std::string>& words) {
        std::vector<int> res;
        std::unordered_map<std::string, int> str2num;
        if (int(s.size()) == 0 || int(words.size()) == 0 || int(words[0].size()) == 0)
            return res;
        for (int i = 0; i < int(words.size()); i++) {
            if (str2num.find(words[i]) == str2num.end())
                str2num.insert(std::unordered_map<std::string, int>::value_type(words[i], 1));
            else
                str2num[words[i]]++;
        }
        for (int i = 0; i <= int(s.size()) - int(words.size()) * int(words[0].size()); i++) {
            std::unordered_map<std::string, int> seen = str2num;
            bool tag = true;
            for (int j = 0; tag && j < int(words.size()); j++) {
                std::string str = s.substr(i + j * words[0].size(), words[0].size());
                if (seen.find(str) != seen.end() && seen[str] > 0) {
                    seen[str]--;
                } else {
                    tag = false;
                }
            }
            if (tag)
                res.push_back(i);
        }
        return res;
    }
};